File: tests/generic/gen_cor_071/Input/event_listener.py
from web3 import Web3
import requests, asyncio
import argparse, json, sys
from eth_account.messages import encode_defunct


def generate_viewing_key(web3, url, private_key):
    sys.stdout.write('Generating viewing key for %s\n' % private_key)
    sys.stdout.flush()

    account = web3.eth.account.privateKeyToAccount(private_key)

    headers = {'Accept': 'application/json', 'Content-Type': 'application/json'}
    data = {"address": account.address}
    response = requests.post('%s/generateviewingkey/' % url, data=json.dumps(data), headers=headers)

    signed_msg = web3.eth.account.sign_message(encode_defunct(text='vk' + response.text), private_key=private_key)
    data = {"signature": signed_msg.signature.hex(), "address": account.address}
    requests.post('%s/submitviewingkey/' % url, data=json.dumps(data), headers=headers)


def handle_event(event):
    sys.stdout.write('%s\n' % Web3.toJSON(event))
    sys.stdout.flush()


async def log_loop(event_filter, poll_interval):
    while True:
        for Stored in event_filter.get_new_entries():
            handle_event(Stored)
        await asyncio.sleep(poll_interval)


def main(contract):
    sys.stdout.write('Starting to run the event loop\n')
    sys.stdout.flush()
    event_filter = contract.events.Stored.createFilter(fromBlock='latest')
    loop = asyncio.get_event_loop()
    try:
        loop.run_until_complete(
            asyncio.gather(
                log_loop(event_filter, 2)))
    finally:
        loop.close()


if __name__ == "__main__":
    parser = argparse.ArgumentParser(prog='event_listener')
    parser.add_argument("url", help="Directory containing the API")
    parser.add_argument("address", help="Address of the spatial engine")
    parser.add_argument("abi", help="Port of the spatial engine")
    parser.add_argument("--pk", help="Private key to register for a viewing key (obscuro only)")
    args = parser.parse_args()

    sys.stdout.write('URL: %s\n' % args.url)
    sys.stdout.write('ADR: %s\n' % args.address)
    sys.stdout.write('ABI: %s\n' % args.abi)
    sys.stdout.flush()

    web3 = Web3(Web3.HTTPProvider(args.url))
    if args.pk: generate_viewing_key(web3, args.url, args.pk)
    with open(args.abi) as f:
        contract = web3.eth.contract(address=args.address, abi=json.load(f))

    main(contract)

File: tests/generic/gen_cor_073/Input/event_listener.js
const Web3 = require('web3');
const commander = require('commander');
const fs = require('fs');

require('console-stamp')(console, 'HH:MM:ss');

function task(contract, from) {
  setTimeout(function() {
    contract.getPastEvents('Stored', { fromBlock: from, toBlock: 'latest'})
    .then(function(events) {
        if (events.length) {
            for (var i = 0, len = events.length; i < len; i+=1) {
                console.log('Stored value =', events[i].returnValues['value']);
                from = events[i].blockNumber+1
            }
        }
        task(contract, from)
    });
  }, 2000);
}

function generate_viewing_key() {
  console.log('Generating viewing key for', options.pk)
  console.log(options.url_http + '/generateviewingkey/')

  fetch(options.url_http +'/generateviewingkey/', {
    method: 'POST',
    headers: {'Accept': 'application/json', 'Content-Type': 'application/json'},
    body: JSON.stringify({address: account.address})
  })
  .then(response => response.text())
  .then((response) => {
         sign_viewing_key()
   })
}

function sign_viewing_key() {
  console.log('Signing viewing key for', options.pk)
  console.log('Result was', response)
  signed_msg = web3.eth.accounts.sign('vk' + response, '0x' + options.pk)

  fetch(options.url_http + '/submitviewingkey/', {
    method: 'POST',
    headers: {'Accept': 'application/json', 'Content-Type': 'application/json'},
    body: JSON.stringify( {signature: signed_msg.signature, address: account.address})
  })
  .then(response => response.text())
  .then((response) => {
    console.log('Starting task ...')
    task(0)
   })
}

commander
  .version('1.0.0', '-v, --version')
  .usage('[OPTIONS]...')
  .option('-u, --url_http <url>', 'Http connection URL')
  .option('-w, --url_ws <url>', 'Web socket connection URL')
  .option('-a, --address <value>', 'Address of the contract')
  .option('-b, --abi <value>', 'Abi of the contract')
  .option('-p, --pk <value>', 'Private key for this client')
  .option('-o, --obscuro', 'True if running against obscuro', false)
  .parse(process.argv);

const options = commander.opts();
console.log('HTTP URL:', `${options.url_http}`);
console.log('WS URL:', `${options.url_ws}`);
console.log('ADR:', `${options.address}`);
console.log('ABI:', `${options.abi}`);
console.log('PK:', `${options.pk}`);
console.log('OB:', `${options.obscuro}`);

var json = fs.readFileSync(`${options.abi}`);
var abi = JSON.parse(json);

const web3 = new Web3(`${options.url_ws}`);
const contract = new web3.eth.Contract(abi, `${options.address}`)
account = web3.eth.accounts.privateKeyToAccount(`${options.pk}`)

if (options.obscuro == true) {
  generate_viewing_key()
}
else {
  console.log('Starting task ...')
  task(contract, 0)
}



